File: BOJ/25_최단경로/1753_최단경로.py
import sys
import heapq

read = sys.stdin.readline

def bfs(st) :
    heap = []
    heapq.heappush(heap, (0,st))
    dist[st] = 0

    while heap :
        cost,cur = heapq.heappop(heap)
        print(cost,cur)
        if dist[cur] < cost : continue

        for nx,nxc in adjmat[cur] :
            nextCost = nxc + cost
            if dist[nx] > nextCost :
                dist[nx] = nextCost
                heapq.heappush(heap, (nextCost,nx))

V,E = map(int,read().split())
K = int(read())
adjmat = [[] for _ in range(V+1)]
INF = sys.maxsize
dist = [INF for _ in range(V+1)]

for _ in range(E) :
    x,y,c = map(int,read().split())
    adjmat[x].append((y,c))

bfs(K)

for i in range(1,V+1) :
    if dist[i] == INF :
        print("INF")
    else :
        print(dist[i])

File: BOJ/25_최단경로/1753_최단경로.cpp
#include <cstdio>
#include <iostream>
#include <queue>
#include <vector>

#define MAX 20000
#define INF 987654321
using namespace std;

int V,E,K,x,y,c;
vector<pair<int,int> > adjmat[MAX+1];
vector<int> dist(MAX+1,INF);

void dji(int st) {

    priority_queue<pair<int,int> > pq; // -cost,cur
    pq.push(make_pair(0,st));
    dist[st] = 0;

    while(!pq.empty()) {
        int cost = -pq.top().first;
        int cur = pq.top().second;
        pq.pop();
        if (dist[cur] < cost) continue;

        for(int i=0; i<adjmat[cur].size(); i++) {
            int next = adjmat[cur][i].first;
            int nextCost = cost + adjmat[cur][i].second;

            if(dist[next] > nextCost) {
                dist[next] = nextCost;
                pq.push(make_pair(-nextCost,next));
            }
        }
    }
}

int main() {

    scanf("%d %d",&V,&E);
    scanf("%d",&K);

    for(int i=0; i<E; i++) {
        scanf("%d %d %d",&x,&y,&c);
        adjmat[x].push_back(make_pair(y,c));
    }
    dji(K);

    for(int i=1; i<=V; i++) {
        if(dist[i] == INF) {
            printf("INF\n");
        }
        else {
            printf("%d\n",dist[i]);
        }
    }

}  